### ISSUE TITLE
Trying to get a value from a Trix editor field

### ISSUE THREAD

**Issue by tanthammar** [livewire/livewire#921]
"livewire/livewire": "^1.0.12",

I am trying to integrate the Trix editor with Livewire.

I add `wire:model` to the hidden trix input field, but the value is not persisted to the Livewire property.

I removed the `type="hidden"` so I can visually see that the form input is updated when I add something to the Trix editor.

Could this have something to do with that the Trix input is populated with html?

```html
@push('body-start')
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/trix/1.2.0/trix.css">
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/trix/1.2.0/trix.js"></script>
@endpush

<input 
    class="trix-input-generator" 
    wire:model="description" 
    id="descr" name="descr" 
    value='{{ $description }}' 
/>
<trix-editor input="descr"></trix-editor>
```

https://github.com/basecamp/trix

**Comment by tanthammar** [comment]
It turns out that this wasn't as easy as I thought.
Apart from not being able to retreive the value from the input, the Trix editor is destroyed on Livewire re-rendering the form on input events.

I got that part solved by re-initialising Trix on livewire hook afterDomUpdate. but I still haven't figured out how to get livewire to persist the input data.

**Comment by devcircus** [comment]
I fought Trix for a long time before finally getting Quill to work. Now that I have Quill working, I could probably go back and get Trix working. However, I haven't been able to make it work short of making the editor a vue component and then passing the value back to livewire via wire:model. 

The main problem I ran into was when editing existing html content (for instance, editing a blog post). wire:model doesn't pass the value down to the component in time to be picked up when the vue component is mounting. nextTick doesn't do the trick either. I had to set up a watcher to get the initial value from livewire. After that, the watcher is ignored.

I think I'll work on it again this evening and try to get it working without making it a vue component.

**Comment by tanthammar** [comment]
I have managed to pass the value to Trix. But I have to init it via JS because it gets destroyed when Livewire updates the component.

This is what I have. ( I have multiple trix fields in the same form, hence the foreach loop in js)

Can you please share how you passed the value to Livewire as I have that problem.
Or, I don't mind usin Quill.
If you can share your solution for that ...

```html
@push('body-start')
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/trix/1.2.0/trix.css">
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/trix/1.2.0/trix.js"></script>
@endpush
```

```html
<input 
wire:model="descriptionSv" 
class="trix-input-generator"
id="descriptionSv" 
name="descriptionSv" 
value='{{ $descriptionSv }}' 
type="hidden" />

<input 
class="trix-input-generator" 
wire:model="descriptionEn" 
id="descriptionEn" 
name="descriptionEn" 
value='{{ $descriptionEn }}' 
type="hidden" />
```

```html
@push('body-end')
<script>
    function initTrix(trixField) {
        let fieldId = trixField.attributes["id"].value;
        let trixEditor = document.createElement("trix-editor");
        trixEditor.setAttribute("input", fieldId);
        let parentDiv = trixField.parentNode;
        parentDiv.insertBefore(trixEditor, trixField);
    }
    function collectTrixFields() {
        let trixfields = document.getElementsByClassName("trix-input-generator");
        for (var i = 0; i < trixfields.length; i++) { 
            initTrix(trixfields.item(i));
        }
    }
    document.addEventListener("turbolinks:load", () => {
        collectTrixFields();
    });
    window.livewire.hook('afterDomUpdate', () => {
        collectTrixFields();
    });
</script>
@endpush
```

**Comment by tanthammar** [comment]
Now I am trying to set the value via Alpinejs, but only the first `@this.set` is used. The second is ignored.

```html
<div x-data="trix()">
<input 
    x-ref="descriptionSv"
    class="trix-input-generator"
    id="descriptionSv" 
    name="descriptionSv" 
    value='{{ $descriptionSv }}' 
    type="hidden" />

<input 
    x-ref="descriptionEn"
    class="trix-input-generator" 
    id="descriptionEn" 
    name="descriptionEn" 
    value='{{ $descriptionEn }}' 
    type="hidden" />

<button x-on:click="update">Set Livewire values</button>

</div>
```

```js
<script>
    function trix() {
        return {
            update() {
                @this.set('descriptionEn', this.$refs.descriptionEn.value);
                @this.set('descriptionSv', this.$refs.descriptionSV.value);
            }
        }
    }
</script>
```

**Comment by devcircus** [comment]
Never tried the editors-in-a-loop. If you make progress, please update this thread for others to see. 

Also, Caleb mentioned on Twitter a while back that he may touch on wysiwygs in his Screencasts.

**Comment by colinmac17** [comment]
@devcircus @tanthammar I am trying to use quill JS

`
@push('scripts')
<script>
    let toolbarOptions = [
        ['bold', 'italic', 'underline', 'link', { 'list': 'ordered'}, { 'list': 'bullet' }]
    ]
    const quill = new Quill('#editor', {
        theme: 'snow',
        modules: {
            toolbar: '#toolbar'
        }
    });

    quill.on('text-change', function(){
        @this.set('body', quill.root.innerHTML);
    })

</script>`

Keep running into `
Unexpected token < in JSON at position 0`

@devcircus do you mind sharing how you got quill js to work? Would love to learn how I can fix my current issues getting it implemented.

**Comment by devcircus** [comment]
When I get to a computer I will. 

Summary:
I made a vue component that handled all of the quill specifics and just listened for the input event in livewire. Inside my vue component, I listen for the editor's change event, grab the html, then emit an input event with that html, which is caught in livewire.

**Comment by colinmac17** [comment]
@devcircus thank you so much! If I figure anything out before I will let you know. I am not using Vue but I still think the approach would be helpful to me.

**Comment by tanthammar** [comment]
I added wire:ignore to the trix container so it is not destroyed everytime Livewire redraws the page.
I still haven't solved how to get the value from an input that is populated with data from trix editor.

**Comment by tanthammar** [comment]
I solved this by collecting formdata with Alpine and emit it to Livewire on save().

**Comment by devcircus** [comment]
Glad you got this worked out for now. The quill editor emits a text-change event every time the contents of the editor changes. I listen for that and then grab the innerHTML, sending it up to the livewire model. 

Hopefully Caleb will touch on this in a video soon.

**Comment by devcircus** [comment]
What I actually did was pull the [editor out of the wink project](https://github.com/writingink/wink/blob/master/resources/js/components/Editor.vue) by @themsaid. I moved some things around and changed what I needed for my project.

Inside the Editor.vue, he [listens for the quill text-change event](https://github.com/writingink/wink/blob/master/resources/js/components/Editor.vue#L83) and emits an input event with the contents of the editor. It's this event that is caught by livewire's ```wire:model``` and keeps the value synced.

The main change I had to make was when on the 'edit' page of a 'post', the editor needs to be populated with existing html from the database. For some reason, when the editor component is mounted, the value hasn't been received from livewire yet. Even if I try to use ```$nextTick```, the value still isn't available. So I set up a watcher in vue on the value, and when it changes, I grab that value and plug it into the editor. After this initial population is done, the watcher can be removed/ignored because any further updates will be from the editor itself, not the backend.
